batch/batch/front_end/query.py

```diff
@@ -0,0 +1,591 @@
+
+    q = request.query.get('q', '')
+    terms = q.split()
+    for _t in terms:
+        if _t[0] == '!':
+            negate = True
+            t = _t[1:]
+        else:
+            negate = False
+            t = _t
+
+        if '=' in t:
+            k, v = t.split('=', 1)
+            if k == 'job_id':
+                condition = '(jobs.job_id = %s)'
+                args = [v]
+            else:
+                condition = '''
+((jobs.batch_id, jobs.job_id) IN
+ (SELECT batch_id, job_id FROM job_attributes
+  WHERE `key` = %s AND `value` = %s))
+'''
+                args = [k, v]
+        elif t.startswith('has:'):
+            k = t[4:]
+            condition = '''
+((jobs.batch_id, jobs.job_id) IN
+ (SELECT batch_id, job_id FROM job_attributes
+  WHERE `key` = %s))
+'''
+            args = [k]
+        elif t in state_search_term_to_states:
+            values = state_search_term_to_states[t]
+            condition = ' OR '.join(['(jobs.state = %s)' for v in values])
+            condition = f'({condition})'
+            args = values
+        else:
+            raise QueryError(f'Invalid search term: {t}.')
+
+        if negate:
+            condition = f'(NOT {condition})'
+
+        where_conditions.append(condition)
+        where_args.extend(args)
+
+    sql = f'''
+WITH base_t AS
+(
+  SELECT jobs.*, batches.user, batches.billing_project, batches.format_version,
+    job_attributes.value AS name
+  FROM jobs
+  INNER JOIN batches ON jobs.batch_id = batches.id
+  INNER JOIN batch_updates ON jobs.batch_id = batch_updates.batch_id AND jobs.update_id = batch_updates.update_id
+  LEFT JOIN job_attributes
+  ON jobs.batch_id = job_attributes.batch_id AND
+    jobs.job_id = job_attributes.job_id AND
+    job_attributes.`key` = 'name'
+  WHERE {' AND '.join(where_conditions)}
+  LIMIT 50
+)
+SELECT base_t.*, COALESCE(SUM(`usage` * rate), 0) AS cost
+FROM base_t
+LEFT JOIN (
+  SELECT aggregated_job_resources_v2.batch_id, aggregated_job_resources_v2.job_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+  FROM base_t
+  LEFT JOIN aggregated_job_resources_v2 ON base_t.batch_id = aggregated_job_resources_v2.batch_id AND base_t.job_id = aggregated_job_resources_v2.job_id
+  GROUP BY aggregated_job_resources_v2.batch_id, aggregated_job_resources_v2.job_id, aggregated_job_resources_v2.resource_id
+) AS usage_t ON base_t.batch_id = usage_t.batch_id AND base_t.job_id = usage_t.job_id
+LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
+GROUP BY base_t.batch_id, base_t.job_id;
+'''
+    sql_args = where_args
+
+    return (sql, sql_args)
+
+
+def parse_batch_jobs_query_v2(request, batch_id):
+    queries: List[Query] = []
+    q = request.query.get('q', '')
+
+    # logic to make time interval queries fast
+    min_start_gt_query: Optional[StartTimeQuery] = None
+    max_end_lt_query: Optional[EndTimeQuery] = None
+
+    if q:
+        terms = q.lstrip().rstrip().split('\n')
+        for term in terms:
```

Comment:
I think that, both for implementation and for documentation, this code could really use a comment with a [BNF grammar](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form). I think the classes you've already defined map to such a grammar very well, so it is mainly useful as a reference and a way to structure the parsing code. While perhaps not significant in practice, this approach spends O(n * m) time parsing query lines (where n is the number of operators and m is the number of query types). I think we can restructure things to take less of a trial-and-error approach that should also hopefully result in fewer lines of code. For example an incomplete grammar might look like:

```
<jobs-query-list> ::= <jobs-query> | <jobs-query> "\n" <jobs-query>

<jobs-query> ::= <cost-query> | <duration-query> | …

<cost-query> ::= "cost" <comparison-operator> <float>
… other query rules …

<comparison-operator> ::= "=" | "==" | "!=" | "<" | …
```

If this were the full grammar, and I'm given the text "cost < 1.0", I don't need to test all the query types to know how to parse this line. I can do the following:

```python
split = line.split()
if len(split) < 3:
    raise QueryError(…)
left, operator, right = split
if left == 'cost':
     # Will throw a QueryError if operator can't parse to a ComparisonOperator or right doesn't parse to a float.
     # It can also produce very precise error messages in those cases.
    queries.append(CostQuery.parse(operator, right))
elif … more rules that map to specific <*-query> rules …
```